File: python/ProgenyImputedFamily.py
from __future__ import annotations

# coding: utf-8
# ProgenyImputedFamily.py
# 片親がknownで後代が一つ以上imputeされている家系を補完する

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import *
from VCFProgenyImputed import *
from Map import *
from KnownFamily import KnownFamily
from OptionSmall import OptionSmall


#################### ProgenyImputedFamily ####################

def impute(orig_vcf: VCFSmall, merged_vcf: VCFGenoBase,
			families: list[KnownFamily], imputed_progenies: list[list[str]],
			ref_haps: list[list[int]], op: OptionSmall) -> Optional[VCFGeno]:
	vcfs: list[VCFGenoBase] = []
	for i in range(len(families)):
		family = families[i]
		progeny = imputed_progenies[i][0]
		samples = [family.mat, family.pat, progeny]
		vcf = VCFFamily.create_by_two_vcfs(merged_vcf, orig_vcf, samples)
		vcf1 = VCFProgenyImputed(samples, vcf.records, ref_haps,
									family.mat_known, op.map, orig_vcf)
		vcf1.impute()
		# 補完した親だけのVCFにする
		parent = family.mat if family.mat_known else family.pat
		vcf_parent = VCFGenoBase.extract_by_samples(vcf1, [parent])
		vcfs.append(vcf_parent)
	
	if not vcfs:
		return None
	
	print("%d families whose progeny is imputed have been imputed." %
															len(families))
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute']

File: src/ProgenyImputedFamily.cpp
#include <algorithm>
#include <cassert>
#include "../include/ProgenyImputedFamily.h"
#include "../include/VCFFamily.h"
#include "../include/VCFProgenyImputed.h"
#include "../include/KnownFamily.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"

using namespace std;


//////////////////// ProgenyImputedFamily ////////////////////

VCFGeno *ProgenyImputedFamily::impute(const VCFSmall *orig_vcf,
									const VCFGenoBase *imputed_vcf,
									const vector<const KnownFamily *>& families,
									const vector<STRVEC>& imputed_progenies,
									const vector<vector<int>>& ref_haps,
									const OptionSmall& op) {
	vector<VCFImputable *>	vcfs;
	vector<VCFFamily *>	vcf_garbage;
	const size_t	L = families.size();
	for(size_t i = 0; i < L; ++i) {
		const KnownFamily	*family = families[i];
		const string&	progeny = imputed_progenies[i][0];
		STRVEC	samples = { family->get_mat(), family->get_pat(), progeny };
		auto	*vcf = VCFFamily::create_by_two_vcfs(imputed_vcf, orig_vcf,
																	samples);
		auto	*vcf1 = new VCFProgenyImputed(samples,
											  vcf->get_family_records(),
											  ref_haps, family->is_mat_known(),
											  op.map, 0.01, orig_vcf);
		vcfs.push_back(vcf1);
		vcf->clear_records();
		delete vcf;
	}
	
	if(vcfs.empty()) {
		Common::delete_all(vcf_garbage);
		return NULL;
	}
	
	// Small VCFs are heavy to process, so it will be parallelized.
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	cout << vcfs.size()
			<< " families whose progeny is imputed have been imputed." << endl;
	
	auto	*new_vcf = VCFImputable::join(vcfs, orig_vcf->get_samples());
	Common::delete_all(vcf_garbage);
	Common::delete_all(vcfs);
	return new_vcf;
}
